dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/smallrye/reactive/messaging/providers/extension/AbstractEmitter.java

```diff
@@ -0,0 +1,210 @@
+package io.smallrye.reactive.messaging.providers.extension;
+
+import static io.smallrye.reactive.messaging.providers.i18n.ProviderExceptions.ex;
+
+import java.util.concurrent.Flow.Publisher;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.messaging.OnOverflow;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.subscription.BackPressureStrategy;
+import io.smallrye.mutiny.subscription.MultiEmitter;
+import io.smallrye.reactive.messaging.EmitterConfiguration;
+import io.smallrye.reactive.messaging.MessagePublisherProvider;
+import io.smallrye.reactive.messaging.providers.helpers.BroadcastHelper;
+import io.smallrye.reactive.messaging.providers.helpers.NoStackTraceException;
+import io.smallrye.reactive.messaging.providers.metrics.MetricDecorator;
+import io.smallrye.reactive.messaging.providers.metrics.MicrometerDecorator;
+import jakarta.enterprise.inject.Instance;
+import jakarta.enterprise.inject.spi.CDI;
+
+public abstract class AbstractEmitter<T> implements MessagePublisherProvider<T> {
+    public static final NoStackTraceException NO_SUBSCRIBER_EXCEPTION = new NoStackTraceException("Unable to process message - no subscriber");
+    protected final AtomicReference<MultiEmitter<? super Message<? extends T>>> internal = new AtomicReference<>();
+    protected final Multi<Message<? extends T>> publisher;
```

Comment:
This class, along with any others taken from an external open source project, should probably have a header indicating its license and where it came from.

I'm unsure if it's _strictly_ needed under the terms of the Apache 2.0 license since the original file has no copyright notice, we haven't made any modifications and we should already include the smallrye RM library in our NOTICES.